src/lib/fingerprint.rs

```diff
@@ -0,0 +1,235 @@
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+
+/// This module implements the algorithm described in M. Chilowicz, E. Duris, G.
+/// Roussel (2009) Syntax Tree Fingerprinting: A Foundation For Source Code
+/// Similarity Detection.
+
+use std::collections::hash_map::DefaultHasher;
+use std::fmt::Debug;
+use std::hash::{Hash, Hasher};
+
+use crypto::digest::Digest;
+use crypto::md5::Md5;
+
+use ast::{Arena, FromOrToNodeId, NodeId};
+use matchers::MappingStore;
+
+const BASE: i64 = 33;
+
+/// Symbol used in calculating static hashes of AST nodes.
+///
+/// See `ast:: to_static_hash_string`.
+pub const OPEN_SYMBOL: &str = "[(";
+
+/// Symbol used in calculating static hashes of AST nodes.
+///
+/// See `ast:: to_static_hash_string`.
+pub const CLOSE_SYMBOL: &str = ")]";
+
+/// Symbol used in calculating static hashes of AST nodes.
+///
+/// See `ast:: to_static_hash_string`.
+pub const SEPARATE_SYMBOL: &str = "@@";
+
+fn calculate_hash<T: Hash>(t: &T) -> i64 {
+    let mut s = DefaultHasher::new();
+    t.hash(&mut s);
+    s.finish() as i64
+}
+
+fn standard_hash<T: Clone + Eq + Hash, U: PartialEq + Copy>(id: &NodeId<U>, arena: &Arena<T, U>) -> i64 {
+    calculate_hash(&arena[*id].ty) + BASE + calculate_hash(&arena[*id].label)
+}
+
+fn in_seed<T: Clone + Eq + Hash + ToString, U: PartialEq + Copy>(id: &NodeId<U>,
+                                                                 arena: &Arena<T, U>)
+                                                                 -> String {
+    String::from(OPEN_SYMBOL) + &arena[*id].label + SEPARATE_SYMBOL + &arena[*id].ty.to_string()
+}
+
+fn out_seed<T: Clone + Eq + Hash + ToString, U: PartialEq + Copy>(id: &NodeId<U>,
+                                                                  arena: &Arena<T, U>)
+                                                                  -> String {
+    arena[*id].ty.to_string() + SEPARATE_SYMBOL + &arena[*id].label + CLOSE_SYMBOL
+}
+
+/// Convert a byte array (from digest) to integer.
+fn bytes_to_int(b: &[u8]) -> i64 {
```

Comment:
`u8` -> `i64` feels a bit weird. I wonder if this should be `u64`?

Reply:
Quite possibly, this is a Java `int` in the GT code